### ISSUE TITLE
ENH: `spsolve` should dispatch to `spsolve_triangular` when triangular matrix is supplied

### ISSUE THREAD

**Issue by dschmitz89** [scipy/scipy#21679]
### Is your feature request related to a problem? Please describe.

Recently, the dense `linalg.solve` function was updated to exploit the matrix structure in https://github.com/scipy/scipy/pull/12824.

The same applies to sparse linear systems: if a triangular matrix is given, a triangular solver is much faster (since https://github.com/scipy/scipy/pull/17924).

### Describe the solution you'd like.

We would need a performant to check if a sparse array is triangular. If, then we should delegate to the triangular solver.

**Comment by dschmitz89** [comment]
CC @perimosocordiae @maxaehle @ilayn : would you have time to comment on this one? The crucial part would be how to do a performant check for triangularity for sparse arrays. I am not sure if we need it for all sparse formats as `spsolve` raises an efficiency warning for some IIRC.

**Comment by nickodell** [comment]
Can we require canonical format for this dispatching? If so, then implementing a check for CSR/CSC can be done by only examining one value per row. It is sufficient to check that the first/last element of each row is within the triangle, depending on whether one is looking for the upper/lower triangle in a CSR/CSC matrix.

**Comment by perimosocordiae** [comment]
Yeah, this should be quite fast for CSR/CSC with sorted indices. We may need to add a native function in sparsetools, but that's not a problem.

**Comment by dschult** [comment]
Do we want a check for triangularity, or a return value stating upper vs lower (thus none or both)

**Comment by bentrainer** [comment]
Is is possible to use [mkl_sparse_?_trsm](https://www.intel.com/content/www/us/en/docs/onemkl/developer-reference-c/2024-2/mkl-sparse-trsm.html) instead of SuperLU's functions when scipy uses MKL as backend?
I found that MATLAB's triangular solver is much faster than `spsolve`, `spsolve_triangular`, and `Factor.solve` from `scikit-sparse` and I don't know if it's because MATLAB use a different BLAS routine...

**Comment by ilayn** [comment]
Matlab uses Tim Davis' SuiteSparse magic for sparse structure discovery and solvers and that's where the speed is coming from. 

We are now doing a structure discovery in dense `linalg.solve` but the main requirement is that if $t_c$ is the structure check time, $t_g$ unstructured solve time, $t_s$ structured solve time, then $t_c$ should be minimal so that the worst case (unstructured) solve time is not too different such that we don't pay for a feature that we do not benefit from that is $t_c \ll t_g$. Simultaneously, $t_s + t_c \ll t_g$ such that we benefit sufficiently that justifies all the work.

There are a lot of options to play around with the permutations to achieve the tightest packing on the diagonal (though the problem is NP-hard). The permuted triangle solve is also quite beneficial in this regard. Tim Davis leave it as an exercise in his book but in fact it is beneficial for sparse arrays to identify the triangular structure. At the end of [matlab's mldivide](https://nl.mathworks.com/help/matlab/ref/double.mldivide.html) there is a flow chart and the references.

**Comment by bentrainer** [comment]
Wow it's surprising to see we could use $\LaTeX$ in the comments!

Yes, Matlab uses SuiteSparse magic, and as for my specific using case of solving $Ax=b$, it utilizes the Cholesky solver, which basically calculates $A = L L^H = R^HR$ first and uses $x=R\backslash R^H\backslash b = L^H\backslash L\backslash b$ to get the solution $x$. This is performed all by CHOLMOD from SuiteSparse in Matlab.

I could also reproduce Matlab's performance using `scikit-sparse`, however I found that CHOLMOD's simplicial routine seems to be single threaded, and if I explicitly use supernodal routine, I could gain ~10% speedup in `scikit-sparse`. Then I tried similar thing in Matlab, it ended up with ~3x faster, which totally confused me. I put a sample in the [gist](https://gist.github.com/bentrainer/ca6a75b39ac3b645423ee767fe2d5307).

In my computer, the results are like:
```bash
# MATLAB
=============================
100 x (A\b) elapsed 2382.13ms
=============================
100 x (R\(R'\b)) elapsed 733.28ms
> err = 3.10e-25
=============================
100 x (L'\(L\b)) elapsed 1218.12ms
> err = 3.10e-25

# Python
spsolve elapsed: 3.70 s
solve_banded elapsed: 3.71 s
sksparse.Factor.solve_A (simplicial) elapsed: 2.52 s
sksparse.Factor.solve_A (supernodal) elapsed: 2.30 s
spsolve_triangular elapsed: 9.32 s
```

Environment:
* i7-12700
* Windows 10
* Matlab 2024b
* Python==3.12, numpy==2.2.0, scipy==1.14.1, scikit-sparse==0.4.14, suitesparse==7.8.2, mkl==2024.1.0


In my naive thought, it is easily multithreaded to solve the problem if `b` is multi-rows. For example, let each thread to solve a single row simultaneously. If so,`spsolve_triangular` should get similar performance compared to Matlab? Noticed that $R\backslash R^H\backslash b$ and $L^H\backslash L\backslash b$ have different performance which may be resulted by memory contiguous.

**Comment by ilayn** [comment]
It is always a bit tricky to make sense of any benchmarks but probably there are some patterns that the libraries are tuned for and maybe they are stated in the documentation to catch the desired performance. I never looked at the code since CHOLMOD and a few other tools are GPLd but I guess there is some time spent on conversions to the desired format.

**Comment by bentrainer** [comment]
Well, this is another sample where I simply implemented a triangular solver with numba parallel, shows under some circumstances (`nrhs` is not small) it could gain noticeable speedup compared to scipy's functions (0.5s vs 0.15s in my env):
https://gist.github.com/bentrainer/ca6a75b39ac3b645423ee767fe2d5307#file-solve_numba-py-L12

I also tested `mkl_sparse_?_trsm` by using `sparse_dot` package in the same [gist](https://gist.github.com/bentrainer/ca6a75b39ac3b645423ee767fe2d5307#file-solve_mkl-py), however it's also single threaded which idk if did anything wrong.

Overall, though I have no idea how to optimize the `scipy.sparse.linalg.spsolve_triangular` now, it might be valuable to think about it.

**Comment by dschmitz89** [comment]
@bentrainer : `spsolve_triangular` simply wraps `SuperLU`, if you have any suggestions that could be filed upstream. From the timings you posted, I would be interested which SciPy version you use. Since 1.14 `spsolve_triangular` uses the triangular solver from `SuperLU`, before it was implemented in Python and a lot slower.

@ilayn : that matlab reference is very interesting. What we have are a general, a triangular and a banded solver from what I understand. What do you think about @nickodell 's [suggestion](https://github.com/scipy/scipy/issues/21679#issuecomment-2406615154) for triangular structure discovery? I would expect it to be reasonably fast.

**Comment by ilayn** [comment]
I'm not up to speed with our current sparse array situation but if there is a way to discover structure, I think that would be fantastic addition. The really cool one would be the permuted triangular solve, but we need to come up with a permuted triangular solver upstream so just take it as a fancy comment. 

By the way, if you consider the image in #12824 and look at the image now in matlab documentation, I think I got lucky selecting the name `bandwidth`, (I don't think we played any role in their choice of words anyways) but it seems like they are doing a similar thing with similar lingo.

**Comment by dschult** [comment]
See #21799 for a performant triangularity checker. 

It uses numpy tools to obtain the rows/cols for all nonzeros and compares them. Shortcuts for full matrices (non-triangular) speed things up for that case. We get the desired checking time much faster than the solving time for either solver. Basically it is in the noise of the solver times. Theoretical scaling supports this result as checking is O(nnz) while solving is O(nnz^2) and O(nnz^3) for the two solvers.
